src/services/identity/NotificationsService.ts

```diff
@@ -0,0 +1,258 @@
+      },
+      order: [
+        ["first_read_time", "DESC NULLS FIRST"],
+        ["priority", "ASC"],
+      ],
+      include: [
+        {
+          model: Site,
+          as: "site",
+          required: true,
+          include: [
+            {
+              model: Repo,
+              required: true,
+              where: {
+                name: siteName,
+              },
+            },
+          ],
+        },
+      ],
+    })
+
+    return this.formatNotifications(notifications)
+  }
+
+  async markNotificationsAsRead({
+    siteName,
+    userId,
+  }: {
+    siteName: string
+    userId: string
+  }) {
+    const siteMember = await this.siteMember.findOne({
+      where: { user_id: userId },
+      include: [
+        {
+          model: Site,
+          required: true,
+          include: [
+            {
+              model: Repo,
+              required: true,
+              where: {
+                name: siteName,
+              },
+            },
+          ],
+        },
+      ],
+    })
+    const readAtDate = new Date()
+    await this.repository.update(
+      {
+        firstReadTime: readAtDate,
+      },
+      {
+        where: {
+          site_member_id: siteMember?.id,
+          first_read_time: null,
+        },
+      }
+    )
+  }
+
+  async create({
+    siteName,
+    userId,
+    link,
+    notificationType,
+    notificationSourceUsername,
+  }: {
+    siteName: string
+    userId: string
+    link: string
+    notificationType: NotificationType
+    notificationSourceUsername: string
+  }) {
+    const siteMember = await this.siteMember.findOne({
+      where: { user_id: userId },
+      include: [
+        {
+          model: Site,
+          required: true,
+          include: [
+            {
+              model: Repo,
+              required: true,
+              where: {
+                name: siteName,
+              },
+            },
+          ],
+        },
+      ],
+    })
+    const recentTargetNotification = await this.repository.findOne({
+      where: {
+        user_id: userId,
+        type: notificationType,
+        created_at: {
+          [Op.gte]: getNotificationExpiryDate(notificationType),
+        },
+        link,
+        source_username: notificationSourceUsername,
+      },
+      include: [
+        {
+          model: Site,
+          as: "site",
+          required: true,
+          include: [
+            {
+              model: Repo,
+              required: true,
+              where: {
+                name: siteName,
+              },
+            },
+          ],
+        },
+      ],
+    })
+
+    if (recentTargetNotification) {
+      // Update existing notification
+      await recentTargetNotification.update({
+        firstReadTime: null,
+        createdAt: new Date(),
+      })
+    } else {
+      // Create new notification
+      await this.repository.create({
+        siteMemberId: siteMember?.id,
+        siteId: siteMember?.siteId,
```

Comment:
Nit: What happens if `siteMember` cannot be found? Would it cause `this.repository.create()` to be called with a `null` value, resulting in a database error?

Reply:
I modified this endpoint in a [later PR](https://github.com/isomerpages/isomercms-backend/pull/523) to use existing entries - this way we can enforce that the user does actually exist

Reply:
Hmm I feel that it might be safer to have an additional check here, but it's just a minor point so feel free to skip!

Reply:
Oh as in the method uses an existing `siteMember` as an argument, so there's no situation where they don't exist